src/platform/pci.zig

```diff
@@ -13,109 +13,124 @@ pub const Handler = struct {
 
 var handlers: [0x100]Handler = undefined;
 
+pub const Cap = struct {
+  addr: Addr,
+  off: u8, 
+  pub fn next(self: *Cap) void { self.off = self.addr.read(u8, self.off + 0x01); }
+  pub fn vndr(self: *const Cap) u8 { return self.addr.read(u8, self.off + 0x00); }
+  pub fn read(self: *const Cap, comptime T: type, off: regoff) T { return self.addr.read(T, self.off + off); }
+  pub fn write(self: *const Cap, comptime T: type, off: regoff, value: T) void { self.addr.write(T, self.off + off, value); }
+};
 
 pub const Addr = struct {
   bus: u8,
   device: u5,
   function: u3,
 
-  pub fn format(self: *const Addr, fmt: []const u8, options: std.fmt.FormatOptions, writer: anytype) !void {
-    try writer.print("[{x:0>2}:{x:0>2}:{x:0>1}]", .{self.bus, self.device, self.function});
+  const vendor_id = cfgreg(u16, 0x00);
+  const device_id = cfgreg(u16, 0x02);
+  const command = cfgreg(u16, 0x04);
+  const status = cfgreg(u16, 0x06);
+  const prog_if = cfgreg(u8, 0x09);
+  const header_type = cfgreg(u8, 0x0E);
+  const base_class = cfgreg(u8, 0x0B);
+  const sub_class = cfgreg(u8, 0x0A);
+  const secondary_bus = cfgreg(u8, 0x19);
+  const cap_ptr = cfgreg(u8, 0x34);
+  const int_line = cfgreg(u8, 0x3C);
+  const int_pin = cfgreg(u8, 0x3D);
+  
+  pub fn cap(self: Addr) Cap { return .{ .addr = self, .off = self.cap_ptr().read() & 0xFC }; }
+
+  pub fn barinfo(self: Addr, bar_idx: u8) BarInfo {
+    var orig: u64 = self.read(u32, 0x10 + bar_idx * 4) & 0xFFFFFFF0;
+    self.write(u32, 0x10 + bar_idx * 4 , 0xFFFFFFFF);
+    var pci_out = self.read(u32, 0x10 + bar_idx * 4);
+    const is64 = ((pci_out & 0b110) >> 1) == 2; // bits 1:2, bar type (0 = 32bit, 1 = 64bit)
+
+    // The BARs can be either 64 or 32 bit, but the trick works for both
+    var response: u64 = @as(u64, pci_out & 0xFFFFFFF0) | 0xFFFFFFFF00000000; 
+    if (is64) {
+      orig |= @as(u64, self.read(u32, 0x14 + bar_idx * 4)) << 32;
+      self.write(u32, 0x10 + bar_idx * 4, 0xFFFFFFFF); // 64bit bar = two 32-bit bars 
+      response |= @as(u64, self.read(u32, 0x14 + bar_idx * 4)) << 32;
+      self.write(u32, 0x14 + bar_idx * 4, @truncate(u32, orig >> 32));
+    }
+    self.write(u32, 0x10 + bar_idx * 4 , @truncate(u32, orig));
+    return .{.phy = orig, .size = ~response + 1};
   }
-};
 
-pub const Device = struct {
-  addr: Addr,
-  vendor_id: u16,
-  device_id: u16,
-  class: u8,
-  subclass: u8,
-  prog_if: u8,
-
-  pub fn format(self: *const Device, fmt: []const u8, options: std.fmt.FormatOptions, writer: anytype) !void {
-    try writer.print("{} ", .{self.addr});
-    try writer.print("{{{x:0>4}:{x:0>4}}} ", .{self.vendor_id, self.device_id});
-    try writer.print("({x:0>2}:{x:0>2}:{x:0>2})", .{self.class, self.subclass, self.prog_if});
+  pub fn read(self: Addr, comptime T: type, offset: regoff) T {
+    if(pci_mmio[self.bus] != null)
+      return std.mem.readIntNative(T, mmio(self, offset)[0..@sizeOf(T)]);
```

Comment:
readIntLittle
Maybe we won't support big-endian now, but it would simplify the lives for us if we will